library/lua/script-data.lua

```diff
@@ -0,0 +1,344 @@
+    ]]
+    --^ Old safety check for before quit_without_saving was added to work around it (the check didn't really work properly anyway)
+    ensure_save_dir_exists("world")
+    path = get_save_dir("world") .. "/" .. id .. ".json"
+  end
+
+  json.encode_file(get_loaded_data(id).data, path)
+end
+
+-- Trigger saves for all persistent tables which have the autosave flag currently set
+-- This is naturally triggered by this script whenever an autosave / quicksave is detected
+function save_autosave()
+  for id, entry in pairs(loaded_data) do
+    if entry.autosave == true then
+      save(id)
+    end
+  end
+end
+
+-- Sets the autosave flag for an entry
+function set_autosave(id, do_autosave)
+  get_loaded_data(id).autosave = do_autosave
+end
+
+-- Set whether the current save should save world entries when quitting
+-- This value is reset back to false whenever a world is unloaded / loaded
+-- It's automatically set to true whenever an arena game is started
+function set_quit_without_saving(bool)
+  quit_without_saving = bool
+end
+
+-- Returns the loaded data entry (essentially the metadata of a loaded script data) for the script data of the given ID
+-- There's not likely a reason to use this externally.
+function get_loaded_data(id)
+  return loaded_data[id]
+end
+
+-- Check to see if save path info is recorded
+-- If it's not, attempt to update the record if possible
+function update_save_path()
+  if recorded_save_path == nil and dfhack.isWorldLoaded() then
+    recorded_save_path = dfhack.getSavePath()
+  end
+end
+---------------------------------------------------------------------
+-- MISC
+
+local function world_load_setup()
+  -- We store the path of the loaded save just in case there's some hiccups with the fact we try to save world data during the SC_WORLD_UNLOADED event (where dfhack.getSavePath() might possibly return nil)
+  recorded_save_path = dfhack.getSavePath()
+
+  -- Begin monitoring for autosaves / quicksaves
+  -- Since autosaves are only actually relevant in fortress mode, there's no reason to start checking for them outside of that mode
+  if dfhack.world.isFortressMode() then
+    start_loop()
+  end
+
+  -- Automatically disable saving on exit if in arena mode
+  -- This is because it's the only current mode(?) that allows for quitting without saving, and there's no way (without some hackery) to really detect that
+  -- DFhack will also report it as having a save directory that doesn't actually exist if the arena game hasn't been saved, just to add to the annoyances
+  if dfhack.world.isArena() then
+    set_quit_without_saving(true)
+  else
+    set_quit_without_saving(false)
+  end
+end
+
+dfhack.onStateChange["script-data"] = function(event)
+  if event == SC_WORLD_LOADED then
+    world_load_setup()
+
+  elseif event == SC_WORLD_UNLOADED then
+    -- Trigger the saves for all `world` entries, provided quitting without saving is disabled
+    if quit_without_saving == false then
```

Comment:
This appears to be the only place you're reading from `quit_without_saving` - could it just be replaced with `isArena()` here? That would avoid the need to keep track of additional global state, if it works.

If you're doing this because isArena() returns false by the time SC_WORLD_UNLOADED fires, maybe it would work under SC_MAP_UNLOADED? Are there any situations where you'd want to support saving data if a map isn't loaded? World generation is the only one I can think of off the top of my head.

In any case, our global "save" event should fire at the right times currently, so #1733 could again make this logic unnecessary (I hope).

Reply:
After a bit of thought, I've remembered that actually both Fortress and Adventure mode can also be exited without saving, so they'd also likely make use of `quit_without_saving`... Next time I get around to editing this current form I'll have to throw something together (either checking `df.global.save_on_exit` that you brought up before, or defaulting to `quit_without_saving` starting off `false` until a map is loaded).

For use, the only thing I can really think of too is worldgen, but I find it doesn't hurt to provide more options to modders in case there's some niche circumstances I didn't foresee.

Though yeah, if/when that event gets added, this'll be a lot more straightforward :P

Reply:
> After a bit of thought, I've remembered that actually both Fortress and Adventure mode can also be exited without saving, so they'd also likely make use of `quit_without_saving`

Oh, are you referring to the ability to abandon a game? Took me a second to think of that too.

I think it would be worth double-checking what our current plugin-facing "save" event does in this case. There is an argument to be made for saving data in this case, even if the user didn't select the "save" option from the menu. After all, fortresses and adventurers remain part of the world (or at least history), even the changes made between the last save and abandoning/retiring.

Reply:
Ah, I was more thinking about how you can load a save after worldgen or losing / abandoning / retiring and go through the world update screen (where time advances by 2 weeks), but then back out without making a fort / adventurer, with none of what happened during the world update being saved. I've always thought it's a bit weird that it works that way, enough so that I often forget it's a thing :b

Reply:
Ah, that's another good case to consider. Pretty sure our plugin-facing save event would not fire in that case. (I'm not actually sure if the "load" event would, though - I'll double-check). I am surprised that loading a fortress or just opening up the embark site selection screen both set the game mode and game type to the same value... I suppose that could throw off some scripts/tools, but they should be able to handle this by only enabling when SC_MAP_LOADED fires or if isMapLoaded() returns true.

Reply:
Update: loading a world from the "start playing" menu *does* fire a "load" event, but the "save" event only fires when the actual DF save screen comes up (or on autosaves). I think the intention is for tools to watch for the "save" event and write any last-minute changes to the persistent tree before it's written to disk. Not seeing any example of plugins that do this currently, but I think this is basically the behavior you're looking to match here, yes?

Reply:
Yep! Saving whenever the game does (be that during save + quit, or autosaves), and skipping saving when the game skips it, is what's desired (for the world-based data, at least).